donkeycar/parts/lidar.py

```diff
@@ -13,6 +14,187 @@
+          
+        if forward_angle < 0 or forward_angle > 360:
+            help.append("forward_angle must be 0 <= forward_angle <= 360")
+            
+        if angle_direction != CLOCKWISE and angle_direction != COUNTER_CLOCKWISE:
+            help.append("angle-direction must be 1 (clockwise) or -1 (counter-clockwise)")
+
+        if len(help) > 0:
+            msg = "Could not start RPLidar; bad parameters passed to constructor"
+            print(msg)
+            for h in help:
+                print("  " + h)
+            raise ValueError(msg)
+
+        self.min_angle = min_angle
+        self.max_angle = max_angle
+        self.min_distance = min_distance
+        self.max_distance = max_distance
+        self.forward_angle = forward_angle
+        self.spin_reverse = (args.angle_direction != CLOCKWISE)
+        self.measurements = [] #a list of (angle, distance, time, scan, index) measurements
+
+        from adafruit_rplidar import RPLidar
+        import glob
+        
+        #
+        # find the serial port where the lidar is connected
+        #
+        port_found = False
+        temp_list = glob.glob ('/dev/ttyUSB*')
+        result = []
+        for a_port in temp_list:
+            try:
+                s = serial.Serial(a_port)
+                s.close()
+                result.append(a_port)
+                port_found = True
+            except serial.SerialException:
+                pass
+        if not port_found:
+            print("No RPLidar is connected.")
+            raise Error("No RPLidar is connected.")
+        
+        # initialize
+        self.port = result[0]
+        self.lidar = RPLidar(None, self.port, timeout=3)
+        self.lidar.clear_input()
+        time.sleep(1)
+        
+        print(self.lidar.info)
+        print(self.lidar.health)
+
+        self.running = True
+
+    def update(self):
+        measurement_count = 0  # number of measurements in the scan
+        measurement_index = 0  # index of next measurement in the scan
+        full_scan_count = 0
+        full_scan_index = 0
+
+        if self.running:
+            try:
+                for new_scan, quality, angle, distance in self.lidar.iter_measurements():
+                    if not self.running:
+                        break
+                                        
+                    now = time.time()
+
+                    # check for start of new scan
+                    if new_scan:
+                        full_scan_count += 1
+                        full_scan_index = 0
+                        measurement_count = measurement_index  # number of points in this full scan
+                        measurement_index = 0   # start filling in next scan
+                        
+                    # rplidar spins clockwise, but we want angles to increase counter-clockwise
+                    if self.spin_reverse:
+                        angle = (360.0 - (angle % 360.0)) % 360.0
```

Comment:
Is this correct? 360.0 - (angle % 360.0) should be in [0.0, 360.0] by construction, why do you still need the % 360.0 afterwards

Reply:
360 - (angle % 360) would turn zero into 360; I want it to be zero.

Reply:
👍